// @paperclipai/shared/dist/environment-support.js
const REMOTE_MANAGED_ADAPTERS = new Set([
    "claude_local",
    "codex_local",
    "cursor",
    "gemini_local",
    "opencode_local",
    "pi_local",
]);
export function adapterSupportsRemoteManagedEnvironments(adapterType) {
    return REMOTE_MANAGED_ADAPTERS.has(adapterType);
}
export function supportedEnvironmentDriversForAdapter(adapterType) {
    return adapterSupportsRemoteManagedEnvironments(adapterType)
        ? ["local", "ssh", "sandbox"]
        : ["local"];
}
export function supportedSandboxProvidersForAdapter(adapterType, additionalProviders = []) {
    return adapterSupportsRemoteManagedEnvironments(adapterType)
        ? Array.from(new Set(additionalProviders))
        : [];
}
export function isEnvironmentDriverSupportedForAdapter(adapterType, driver) {
    return supportedEnvironmentDriversForAdapter(adapterType).includes(driver);
}
export function isSandboxProviderSupportedForAdapter(adapterType, provider, additionalProviders = []) {
    if (!provider)
        return false;
    return supportedSandboxProvidersForAdapter(adapterType, additionalProviders).includes(provider);
}
export function getAdapterEnvironmentSupport(adapterType, additionalSandboxProviders = []) {
    const supportedDrivers = new Set(supportedEnvironmentDriversForAdapter(adapterType));
    const supportedProviders = new Set(supportedSandboxProvidersForAdapter(adapterType, additionalSandboxProviders));
    const sandboxProviders = {
        fake: "unsupported",
    };
    for (const provider of additionalSandboxProviders) {
        sandboxProviders[provider] = supportedProviders.has(provider)
            ? "supported"
            : "unsupported";
    }
    return {
        adapterType,
        drivers: {
            local: supportedDrivers.has("local") ? "supported" : "unsupported",
            ssh: supportedDrivers.has("ssh") ? "supported" : "unsupported",
            sandbox: supportedDrivers.has("sandbox") ? "supported" : "unsupported",
            plugin: supportedDrivers.has("plugin") ? "supported" : "unsupported",
        },
        sandboxProviders,
    };
}
export function getEnvironmentCapabilities(adapterTypes, options = {}) {
    const pluginProviderKeys = Object.keys(options.sandboxProviders ?? {});
    const sandboxProviders = {
        fake: {
            status: "unsupported",
            supportsSavedProbe: true,
            supportsUnsavedProbe: true,
            supportsRunExecution: false,
            supportsReusableLeases: true,
            displayName: "Fake",
            source: "builtin",
        },
    };
    for (const [provider, capability] of Object.entries(options.sandboxProviders ?? {})) {
        sandboxProviders[provider] = {
            status: capability.status ?? "supported",
            supportsSavedProbe: capability.supportsSavedProbe ?? true,
            supportsUnsavedProbe: capability.supportsUnsavedProbe ?? true,
            supportsRunExecution: capability.supportsRunExecution ?? true,
            supportsReusableLeases: capability.supportsReusableLeases ?? true,
            displayName: capability.displayName,
            description: capability.description,
            source: capability.source ?? "plugin",
            pluginKey: capability.pluginKey,
            pluginId: capability.pluginId,
            configSchema: capability.configSchema,
        };
    }
    return {
        adapters: adapterTypes.map((adapterType) => getAdapterEnvironmentSupport(adapterType, pluginProviderKeys)),
        drivers: {
            local: "supported",
            ssh: "supported",
            sandbox: "supported",
            plugin: "unsupported",
        },
        sandboxProviders,
    };
}
